Expand config validation logic and tests

cfunkhouser opened the pull request:
This change adds an extra check for provider validity at config parse time, and expands the tests. There are no multi-provider tests in this change.

cfunkhouser committed:
Expand config validation logic and tests

asjoyner commented:
as per table-side discussion, please do away with the validProviders string slice, and implement the awesome register provider and associated magic.

cfunkhouser committed:
Register drive client providers on module import.

cfunkhouser commented:
Okay, I've implemented and tested the drive client provider registration logic we discussed around the table.

This gets us some more complexity in the drive module; namely:
-   A `RegisterProvider` function, which is meant to be called from the module init() function
-   A `ValidProvider` function which checks if a provider is registered
-   A `NewClient` function - which has the same signature as the per-provider constructor - and creates the new client of the correct provider type based on the `Provider` member in the `Config` struct

I've also updated the drive provider modules, and shadeutil.

cfunkhouser committed:
Clean up some files which missed a gofmt

cfunkhouser committed:
Unexport clientCreator

cfunkhouser committed:
Use a RWMutex for protecting provider map instead of Mutex.
